File: backend/app/services/ai.py
import httpx
import json
import logging
from app.config import AI_CONFIG

logger = logging.getLogger(__name__)

async def call_aliyun(api_key: str, system_prompt: str, user_query: str):
    """Call Alibaba Cloud DashScope API (non-streaming)"""
    logger.info(f"[Aliyun Call] Prompt:\nSystem: {system_prompt}\nUser: {user_query}")
    url = "https://dashscope.aliyuncs.com/api/v1/services/aigc/text-generation/generation"
    
    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            url,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json"
            },
            json={
                "model": AI_CONFIG["aliyun"]["model"],
                "input": {
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_query}
                    ]
                },
                "parameters": {
                    "result_format": "message"
                }
            }
        )
        
        if response.status_code != 200:
            logger.error(f"[Aliyun Call] Error: {response.text}")
            raise Exception(f"Aliyun API error: {response.text}")
        
        data = response.json()
        content = data["output"]["choices"][0]["message"]["content"]
        logger.info(f"[Aliyun Call] Response: {content}")
        return {"content": content}

async def call_google(api_key: str, system_prompt: str, user_query: str):
    """Call Google Gemini API (non-streaming)"""
    model = AI_CONFIG["google"]["model"]
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={api_key}"
    
    combined_prompt = f"{system_prompt}\n\nUser Query: {user_query}"
    
    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            url,
            headers={"Content-Type": "application/json"},
            json={
                "contents": [{
                    "parts": [{"text": combined_prompt}]
                }]
            }
        )
        
        if response.status_code != 200:
            raise Exception(f"Google API error: {response.text}")
        
        data = response.json()
        content = data["candidates"][0]["content"]["parts"][0]["text"]
        return {"content": content}

async def stream_aliyun(api_key: str, system_prompt: str, user_query: str):
    """Stream from Alibaba Cloud DashScope API"""
    logger.info(f"[Aliyun Stream] Prompt: System={system_prompt[:50]}... User={user_query[:50]}...")
    url = "https://dashscope.aliyuncs.com/api/v1/services/aigc/text-generation/generation"
    
    async with httpx.AsyncClient(timeout=120.0) as client:
        async with client.stream(
            "POST",
            url,
            headers={
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "X-DashScope-SSE": "enable"
            },
            json={
                "model": AI_CONFIG["aliyun"]["model"],
                "input": {
                    "messages": [
                        {"role": "system", "content": system_prompt},
                        {"role": "user", "content": user_query}
                    ]
                },
                "parameters": {
                    "result_format": "message",
                    "incremental_output": True
                }
            }
        ) as response:
            if response.status_code != 200:
                error_text = await response.aread()
                logger.error(f"[Aliyun Stream] Error: {error_text.decode()}")
                raise Exception(f"Aliyun API error: {error_text.decode()}")
            
            async for line in response.aiter_lines():
                if line.startswith("data:"):
                    data_str = line[5:].strip()
                    if data_str:
                        try:
                            data = json.loads(data_str)
                            if "output" in data and "choices" in data["output"]:
                                content = data["output"]["choices"][0]["message"].get("content", "")
                                if content:
                                    logger.debug(f"[Aliyun Stream] Chunk: {content}")
                                    yield content
                        except json.JSONDecodeError:
                            pass

async def stream_google(api_key: str, system_prompt: str, user_query: str):
    """Stream from Google Gemini API"""
    model = AI_CONFIG["google"]["model"]
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:streamGenerateContent?key={api_key}"
    
    combined_prompt = f"{system_prompt}\n\nUser Query: {user_query}"
    
    async with httpx.AsyncClient(timeout=120.0) as client:
        async with client.stream(
            "POST",
            url,
            headers={"Content-Type": "application/json"},
            json={
                "contents": [{
                    "parts": [{"text": combined_prompt}]
                }]
            }
        ) as response:
            if response.status_code != 200:
                error_text = await response.aread()
                raise Exception(f"Google API error: {error_text.decode()}")
            
            buffer = ""
            async for chunk in response.aiter_text():
                buffer += chunk
                try:
                    if buffer.strip().startswith("["):
                        buffer = buffer.strip()[1:]
                    
                    parts = buffer.split("},{")
                    for i, part in enumerate(parts[:-1]):
                        if not part.startswith("{"):
                            part = "{" + part
                        if not part.endswith("}"):
                            part = part + "}"
                        try:
                            data = json.loads(part)
                            if "candidates" in data:
                                text = data["candidates"][0]["content"]["parts"][0].get("text", "")
                                if text:
                                    yield text
                        except:
                            pass
                    buffer = parts[-1] if parts else ""
                except:
                    pass

File: archive/js/js/ai.js
import { AI_CONFIG } from './config.js';

/**
 * Unified AI Service to handle requests to different providers.
 */
export class AIService {

    constructor() {
        this.config = AI_CONFIG;
    }

    /**
     * Send a query to the configured AI provider.
     * @param {string} systemPrompt - The system instruction (context, role).
     * @param {string} userQuery - The user's question or input.
     * @returns {Promise<string>} - The AI's response text.
     */
    async chat(systemPrompt, userQuery) {
        const provider = this.config.provider;

        if (provider === 'aliyun') {
            return this._callAliyun(systemPrompt, userQuery);
        } else if (provider === 'google') {
            return this._callGoogle(systemPrompt, userQuery);
        } else {
            throw new Error(`Unknown provider: ${provider}`);
        }
    }

    /**
     * Call Alibaba Cloud DashScope API (Qwen)
     */
    async _callAliyun(systemPrompt, userQuery) {
        const { apiKey, model } = this.config.aliyun;
        if (!apiKey) throw new Error("Aliyun API Key is missing in js/config.js");

        const url = 'https://dashscope.aliyuncs.com/api/v1/services/aigc/text-generation/generation';

        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                model: model,
                input: {
                    messages: [
                        { role: 'system', content: systemPrompt },
                        { role: 'user', content: userQuery }
                    ]
                },
                parameters: {
                    result_format: 'message'
                }
            })
        });

        if (!response.ok) {
            const err = await response.json();
            throw new Error(`Aliyun Error: ${err.message || response.statusText}`);
        }

        const data = await response.json();
        return data.output.choices[0].message.content;
    }

    /**
     * Call Google Gemini API
     */
    async _callGoogle(systemPrompt, userQuery) {
        const { apiKey, model } = this.config.google;
        if (!apiKey) throw new Error("Google API Key is missing in js/config.js");

        const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;

        // Gemini doesn"t verify have a strict "system" role in the 1.0 endpoint same way, 
        // but we can merge it or use the new structure. 
        // For simplicity/compatibility, we concatenate.
        const combinedPrompt = `${systemPrompt}\n\nUser Query: ${userQuery}`;

        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                contents: [{
                    parts: [{
                        text: combinedPrompt
                    }]
                }]
            })
        });

        if (!response.ok) {
            const err = await response.json();
            throw new Error(`Google Error: ${err.error?.message || response.statusText}`);
        }

        const data = await response.json();
        return data.candidates[0].content.parts[0].text;
    }
    /**
     * Analyze text for Reading Coach (Smart Import).
     * Returns structured JSON with segmentation, keywords, and insights.
     */
    async analyzeText(rawText) {
        const systemPrompt = `You are a linguistic engine for an English learning app. 
        Analyze the user's text. 
        1. Split the text into INDIVIDUAL SENTENCES. However, if a sentence is very short (less than 6 words), merge it with the previous or next sentence to keep flow.
        2. Return a VALID JSON object (no markdown formatting, just raw JSON).
        Structure:
        {
          "paragraphs": [
            {
              "text": "Sentence text...",
              "translation": "Full Chinese translation of this paragraph.",
              "insight": { "tag": "Short Theme (e.g. Irony)", "text": "Deep analysis..." },
              "knowledge": [
                { 
                  "key": "unique_id_for_word", 
                  "word": "Display Word", 
                  "ipa": "IPA", 
                  "def": "Chinese Definition", 
                  "clue": "English Hint/Synonym", 
                  "diff": 1-6 (Integer, 1=A1, 6=C2), 
                  "context": "Short sentence context or collocation" 
                }
              ]
            }
          ]
        }
        Do not include \`\`\`json blocks. Just the raw JSON string.`;

        // We use the same chat interface but with a different prompt
        // Note: For large texts, we might need chunking, but for demo we assume reasonable length.
        const responseText = await this.chat(systemPrompt, rawText);

        console.log("Raw AI Response:", responseText); // Debug log

        // Clean up response if it contains markdown code blocks
        // Regex to capture content inside ```json ... ``` or just ``` ... ```
        let jsonStr = responseText;
        const codeBlockMatch = responseText.match(/```(?:json)?\s*([\s\S]*?)\s*```/);
        if (codeBlockMatch) {
            jsonStr = codeBlockMatch[1];
        }

        jsonStr = jsonStr.trim();

        try {
            return JSON.parse(jsonStr);
        } catch (e) {
            console.error("AI JSON Parse Error:", e, jsonStr);
            // Fallback: try to find start and end of JSON array/object
            const firstBrace = jsonStr.indexOf('{');
            const lastBrace = jsonStr.lastIndexOf('}');
            if (firstBrace !== -1 && lastBrace !== -1) {
                try {
                    return JSON.parse(jsonStr.substring(firstBrace, lastBrace + 1));
                } catch (e2) {
                    throw new Error("Failed to parse AI response. " + e.message);
                }
            }
            throw new Error("Failed to parse AI response. " + e.message);
        }
    }
}

export const aiService = new AIService();
